components/butterfly/src/server/pull.rs

```diff
@@ -59,7 +63,9 @@ impl Pull {
             let msg = match socket.recv_msg(0) {
                 Ok(msg) => msg,
                 Err(e) => {
-                    error!("Error receiving message: {:?}", e);
+                    if e != zmq::Error::EAGAIN {
```

Comment:
I think that it _might_ be useful to add a comment here about why we're doing this. If you're not aware of this current deadlock monitoring situation, you could be forgiven for wondering what the heck we're doing here.

Also, somebody that's not catching the fact that this particular channel is a ZMQ one, as opposed to one of the the increasing number of Rust channels in the rest of the codebase might not realize that there's an inherent timeout on this `recv` message.

I know this is planning _way_ ahead, but I see a day in our future when we can drop ZMQ for pure Rust everywhere, and it would be good to make sure that the intention of this change is very clear, so we can faithfully translate it.

Based on my ZMQ knowledge, this looks good. @markan would be a good second opinion to get here to confirm, though.

Reply:
It looks reasonable to me. I assume this means you will be limiting the maximum value for the RecvTimeoutMillis to be less than the maximum heartbeat interval?  The default 1 minute heartbeat check and 5 s Recv timeout should be fine, but might as well block a footgun.

Reply:
oooh, that's a great point @markan

Reply:
I don't think it's worth exposing `ThreadAliveThreshold` (which is private to the `habitat_common::sync` module) and making this more complicated. The defaults should work well and the consequences of setting the timeout too high relative to `ThreadAliveThreshold` would just be some spurious logging.

To account for the case that we may want to disable the receive timeout altogether (by setting it to the default of -1), I'll add a check for that value and opt this thread out of the liveliness checking altogether.

If we run into problems, we can always make this more complex later (perhaps by letting the calling thread specify how long must pass without a heartbeat to consider it dead), but for now I'd like to keep it simple.

Reply:
Does https://github.com/habitat-sh/habitat/pull/6565/commits/83d38218a723c2dc8758817c709dfde6bea7f65c work for y'all?

Reply:
![tenor-242646827](https://user-images.githubusercontent.com/207178/57969833-25c84980-7948-11e9-9a3c-bd2be95ff091.gif)

Ship it!